README.md
# vl53l1x_ros2

This is STM [VL53L1X](https://www.st.com/en/imaging-and-photonics-solutions/vl53l1x.html) time-of-flight rangefinder driver for ROS2. Tested on a Raspberry Pi 3 and 4 with [CJMCU-531](https://ru.aliexpress.com/item/VL53L1X/32911692450.html) board.

[This is in progress]

The code is based on the [STM VL53L1X API library](https://www.st.com/content/st_com/en/products/embedded-software/proximity-sensors-software/stsw-img007.html).

## Connecting

Example of I²C connecting VL53L1X module to Raspberry Pi:

<img src="raspberry-vl53l1x.png" width=500>

## Installation

### Manual

1. Clone the repository into your Colcon workspace:

    ```bash
    cd ~/dev_ws/src
    git clone https://github.com/mhl787156/vl53l1x_ros2.git
    ```

2. Build the package:

    ```bash
    cd ~/dev_ws
    colcon build --packages_select vl53l1x
    ```

## Quick start

Run with the default settings:

```bash
ros2 run vl53l1x vl53l1x_node
```

See the ranging results:

```bash
ros2 topic echo /vl53l1x/range
```

See the ranging rate:

```bash
ros2 topic hz /vl53l1x/range
```

## Parameters

All parameters are optional. Check the [`launch`](https://github.com/okalachev/vl53l1x_ros/tree/master/vl53l1x/launch) folder for launch-file templates.

* `~i2c_bus` (*int*) – I2C bus number (default: 1).
* `~i2c_address` (*int*) – I2C address (default: 41 (i.e. 0x29)). (Note this may need to be given as a base 10 integer within launch files.)
* `~mode` (*int*) – distance mode, 1 = short, 2 = medium, 3 = long (default: 3).
* `~timing_budget` (*double*) – timing budget for measurements, *s* (default: 0.1)
* `~poll_rate` (*double*) – polling data rate, *Hz* (default: 100.0).
* `~ignore_range_status` (*bool*) – ignore validness of measurements (default: false).
* `~min_signal` (*double*) – minimum amplitude of the signal reflected from the target to be considered valid, *MCPS* (default: 1).
* `~max_sigma` (*double*) – maximum standard deviation of the measurement to be considered valid, *m* (default: 0.015).
* `~offset` (*float*) – offset to be automatically added to measurement value, *m* (default: 0.0).
* `~frame_id` (*string*) – frame id for output `Range` messages (default: "").
* `~field_of_view` (*float*) – field of view for output `Range` messages, *rad* (default: 0.471239).
* `~min_range` (*float*) – minimum range for output `Range` messages, *m* (default: 0.0).
* `~max_range` (*float*) – maximum range for `Range` output messages, *m* (default: 4.0).

`timing_budget` is the time VL53L1X uses for ranging. The larger this time, the more accurate is measurement and the larger is maximum distance. Timing budget can be set from *0.02 s* up to *1 s*.

* *0.02 s* is the minimum timing budget and can be used only in *Short* distance mode.
* *0.033 s* is the minimum timing budget which can work for all distance modes.
* *0.14 s* is the timing budget which allows the maximum distance of *4 m* (in *Long* distance mode).

The resulting measurement rate is *1 / (timing budget + 0.004) Hz*.

Setting `ignore_range_status` to `true` makes the node to ignore `RangeStatus` field of measurements. This may significantly improve maximum distance and rate but affects quality of measurements.

`mode` is one of three distance modes, with the timing budget of *0.1 s*, *Short*, *Medium* and *Long* modes have maximum distances of 136, 290, and 360 cm, respectively.

`~min_range` and `~max_range` don't affect any device settings and only define `min_range` and `max_range` values of the output [`Range`](http://docs.ros.org/melodic/api/sensor_msgs/html/msg/Range.html) messages.

See the [official documentation](https://www.st.com/resource/en/datasheet/vl53l1x.pdf) for more information on mode and timing budget.

## Topics

### Published

* `~range` ([*sensor_msgs/Range*](http://docs.ros.org/kinetic/api/sensor_msgs/html/msg/Range.html)) – resulting measurement.
* `~data` ([*vl53l1x/MeasurementData*](https://github.com/mhl787156/vl53l1x_ros2/blob/master/vl53l1x/msg/MeasurementData.msg)) – additional data of the measurement.

## Copyright

Modified 2021 Mickey Li
Copyright © 2019 Oleg Kalachev.

Distributed under BSD 3-Clause License (https://opensource.org/licenses/BSD-3-Clause).


vl53l1x/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package vl53l1x
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.1 (2021-08-17)
------------------
* Ported to ROS2
* CMAKE updated to ament format. Building of lib and i2c moved into new CmakeLists.txt inside /lib
* MeasurementData.msg explicilty uses std_msgs/Header
* i2c_setup now throws runtime_error instead of ROS errors directly for portability.
* Contributors: Mickey Li

1.0.0 (2021-01-28)
------------------
* Change licence to BSD-3-Clause

0.4.1 (2021-01-16)
------------------

0.4.0 (2019-12-05)
------------------
* Publish additional data of the measurements
* Publish measurement data even if range status is not valid
* Add status meanings to measurement data message file
* Implement min_signal and max_sigma parameters
* Add pass_statuses parameter for passing range statuses other than 0
* Print device info on startup
* Contributors: Alexey Rogachevskiy, Arthur Golubtsov, Oleg Kalachev


vl53l1x/lib/i2c/i2c.cpp
#include <stdlib.h>
#include <linux/i2c-dev.h>
#include <sys/ioctl.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <arpa/inet.h>
#include <stdexcept>
#include <memory.h>


#include "i2c.h"

using namespace std;

int fd;

void i2c_setup(uint8_t bus, uint8_t addr)
{
	char filename[50];
	sprintf(filename, "/dev/i2c-%d", bus);

	fd = open(filename, O_RDWR);

	if (fd < 0) {
		char buffer [50];
		sprintf(buffer, "Failed to open I2C bus %d", bus);
		throw std::runtime_error(buffer);
	}

	int rc = ioctl(fd, I2C_SLAVE, addr);
	if (rc < 0) {
  		char buffer [50];
		sprintf(buffer, "Failed to select I2C device with address 0x%02x ", bus);
		throw std::runtime_error(buffer);
	}
}

void i2c_release()
{
	close(fd);
}

bool i2c_readRegisterMulti(uint16_t reg, size_t count, void *pdata)
{
	reg = htons(reg); // little endian to big endian

	// write register number
	if (write(fd, &reg, sizeof(reg)) < 0) {
		return false;
	}

	return read(fd, pdata, count);
}

bool i2c_readRegisterByte(int reg, uint8_t *pdata)
{
	return i2c_readRegisterMulti(reg, 1, pdata) > 0;
}

bool i2c_readRegisterWord(uint16_t reg, uint16_t *pdata)
{
	if (i2c_readRegisterMulti(reg, 2, pdata) < 0) {
		return false;
	}
	*pdata = ntohs(*pdata); // big endian to little endian
	return true;
}

bool i2c_writeRegisterMulti(uint16_t reg, size_t count, void *pdata)
{
	uint8_t buf[count + sizeof(reg)];

	reg = htons(reg); // little endian to big endian

	memcpy(buf, &reg, sizeof(reg));
	memcpy(buf + sizeof(reg), pdata, count);

	return write(fd, buf, count + sizeof(reg)) > 0;
}

bool i2c_writeRegisterByte(uint16_t reg, uint8_t value)
{
	return i2c_writeRegisterMulti(reg, 1, &value);
}


bool i2c_writeRegisterWord(uint16_t reg, uint16_t value)
{
	value = htons(value); // little endian to big endian
	return i2c_writeRegisterMulti(reg, 2, &value);
}


vl53l1x/src/vl53l1x.cpp
/*
 * STM VL53L1X ToF rangefinder driver for ROS
 *
 * Author: Oleg Kalachev <okalachev@gmail.com>
 *
 * Distributed under BSD 3-Clause License (available at https://opensource.org/licenses/BSD-3-Clause).
 *
 * Documentation used:
 * VL53L1X datasheet - https://www.st.com/resource/en/datasheet/vl53l1x.pdf
 * VL53L1X API user manual - https://www.st.com/content/ccc/resource/technical/document/user_manual/group0/98/0d/38/38/5d/84/49/1f/DM00474730/files/DM00474730.pdf/jcr:content/translations/en.DM00474730.pdf
 *
 */

#include <string>
#include <vector>
#include <stdexcept>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/range.hpp>
#include <vl53l1x/msg/measurement_data.hpp>

#include "vl53l1_api.h"
#include "i2c.h"

#define xSTR(x) #x
#define STR(x) xSTR(x)

#define CHECK_STATUS(func) { \
	VL53L1_Error status = func; \
	if (status != VL53L1_ERROR_NONE) { \
		RCLCPP_WARN(this->get_logger(), "VL53L1X: Error %d on %s", status, STR(func)); \
	} \
}

class VL53L1X_Node: public rclcpp::Node {
	public:
		VL53L1X_Node();

	private:
		void poll();

		sensor_msgs::msg::Range range;
		vl53l1x::msg::MeasurementData data;

		int mode, i2c_bus, i2c_address;
		double poll_rate, timing_budget, offset;
		bool ignore_range_status;
		std::vector<int64_t> pass_statuses { VL53L1_RANGESTATUS_RANGE_VALID,
										VL53L1_RANGESTATUS_RANGE_VALID_NO_WRAP_CHECK_FAIL,
										VL53L1_RANGESTATUS_RANGE_VALID_MERGED_PULSE };

		rclcpp::Publisher<sensor_msgs::msg::Range>::SharedPtr range_pub;
		rclcpp::Publisher<vl53l1x::msg::MeasurementData>::SharedPtr data_pub;

		rclcpp::TimerBase::SharedPtr timer;

		VL53L1_Dev_t dev;
		VL53L1_Error dev_error;
};

VL53L1X_Node::VL53L1X_Node() :
	Node("vl53l1X_rangefinder",
		 "",
		 rclcpp::NodeOptions()
			.allow_undeclared_parameters(true)
			.automatically_declare_parameters_from_overrides(true)
	)
{
	this->get_parameter_or("mode", this->mode, 3);
	this->get_parameter_or("i2c_bus", this->i2c_bus, 1);
	this->get_parameter_or("i2c_address", this->i2c_address, 0x29);
	this->get_parameter_or("poll_rate", this->poll_rate, 100.0);
	this->get_parameter_or("ignore_range_status", this->ignore_range_status, false);
	this->get_parameter_or("timing_budget", this->timing_budget, 0.1);
	this->get_parameter_or("offset", this->offset, 0.0);
	this->get_parameter_or("frame_id", this->range.header.frame_id, std::string(""));
	this->get_parameter_or("field_of_view", this->range.field_of_view, 0.471239f); // 27 deg, source: datasheet
	this->get_parameter_or("min_range", this->range.min_range, 0.0f);
	this->get_parameter_or("max_range", this->range.max_range, 4.0f);
	try {
		this->pass_statuses = this->get_parameter("pass_statuses").as_integer_array();
	} catch (const rclcpp::ParameterTypeException& e) {
		RCLCPP_INFO(this->get_logger(), "Pass Statuses set as default.");
	}

	this->range_pub = this->create_publisher<sensor_msgs::msg::Range>("range", 20);
	this->data_pub = this->create_publisher<vl53l1x::msg::MeasurementData>("range_data", 20);

	if (this->timing_budget < 0.02 || this->timing_budget > 1) {
		RCLCPP_FATAL(this->get_logger(), "Error: timing_budget should be within 0.02 and 1 s (%g is set)", this->timing_budget);
		rclcpp::shutdown();
	}

	this->range.radiation_type = sensor_msgs::msg::Range::INFRARED;
	// 	// The minimum inter-measurement period must be longer than the timing budget + 4 ms (*)
	double inter_measurement_period = this->timing_budget + 0.004;

	// Setup I2C bus
	i2c_setup(this->i2c_bus, this->i2c_address);

	// Init sensor
	// VL53L1_Dev_t dev;
	// VL53L1_Error dev_error;
	VL53L1_software_reset(&this->dev);
	VL53L1_WaitDeviceBooted(&this->dev);
	VL53L1_DataInit(&this->dev);
	VL53L1_StaticInit(&this->dev);
	VL53L1_SetPresetMode(&this->dev, VL53L1_PRESETMODE_AUTONOMOUS);

	// Print device info
	VL53L1_DeviceInfo_t device_info;
	CHECK_STATUS(VL53L1_GetDeviceInfo(&this->dev, &device_info));
	RCLCPP_INFO(this->get_logger(), "VL53L1X: Device name: %." STR(VL53L1_DEVINFO_STRLEN) "s", device_info.Name);
	RCLCPP_INFO(this->get_logger(), "VL53L1X: Device type: %." STR(VL53L1_DEVINFO_STRLEN) "s", device_info.Type);
	RCLCPP_INFO(this->get_logger(), "VL53L1X: Product ID: %." STR(VL53L1_DEVINFO_STRLEN) "s", device_info.ProductId);
	RCLCPP_INFO(this->get_logger(), "VL53L1X: Type: %u Version: %u.%u", device_info.ProductType,
	          device_info.ProductRevisionMajor, device_info.ProductRevisionMinor);

	// Setup sensor
	CHECK_STATUS(VL53L1_SetDistanceMode(&this->dev, mode));
	CHECK_STATUS(VL53L1_SetMeasurementTimingBudgetMicroSeconds(&dev, round(timing_budget * 1e6)));

	double min_signal;
	if (this->get_parameter("min_signal", min_signal)) {
		CHECK_STATUS(VL53L1_SetLimitCheckValue(&this->dev, VL53L1_CHECKENABLE_SIGNAL_RATE_FINAL_RANGE, min_signal * 65536));
	}

	double max_sigma;
	if (this->get_parameter("max_sigma", max_sigma)) {
		CHECK_STATUS(VL53L1_SetLimitCheckValue(&this->dev, VL53L1_CHECKENABLE_SIGMA_FINAL_RANGE, max_sigma * 1000 * 65536));
	}

	// Start sensor
	for (int i = 0; i < 100; i++) {
		CHECK_STATUS(VL53L1_SetInterMeasurementPeriodMilliSeconds(&this->dev, round(inter_measurement_period * 1e3)));
		this->dev_error = VL53L1_StartMeasurement(&this->dev);
		if (this->dev_error == VL53L1_ERROR_INVALID_PARAMS) {
			inter_measurement_period += 0.001; // Increase inter_measurement_period to satisfy condition (*)
		} else break;
	}

	// Check for errors after start
	if (this->dev_error != VL53L1_ERROR_NONE) {
		RCLCPP_FATAL(this->get_logger(), "VL53L1X: Can't start measurement: error %d", this->dev_error);
		rclcpp::shutdown();
	}

	RCLCPP_INFO(this->get_logger(), "VL53L1X: ranging");

	this->timer = this->create_wall_timer(
		std::chrono::duration<double>(1.0 / this->poll_rate),
		std::bind(&VL53L1X_Node::poll, this)
	);

}

void VL53L1X_Node::poll() {
	VL53L1_RangingMeasurementData_t measurement_data;
	this->range.header.stamp = this->now();

	// Check the data is ready
	uint8_t data_ready = 0;
	VL53L1_GetMeasurementDataReady(&this->dev, &data_ready);
	if (!data_ready) {
		return;
	}

	// Read measurement
	VL53L1_GetRangingMeasurementData(&this->dev, &measurement_data);
	VL53L1_ClearInterruptAndStartMeasurement(&this->dev);

	// Publish measurement data
	this->data.header.stamp = this->range.header.stamp;
	this->data.signal = measurement_data.SignalRateRtnMegaCps / 65536.0;
	this->data.ambient = measurement_data.AmbientRateRtnMegaCps / 65536.0;
	this->data.effective_spad = measurement_data.EffectiveSpadRtnCount / 256;
	this->data.sigma = measurement_data.SigmaMilliMeter / 65536.0 / 1000.0;
	this->data.status = measurement_data.RangeStatus;
	this->data_pub->publish(this->data);

	// Check measurement for validness
	if (!this->ignore_range_status &&
		std::find(this->pass_statuses.begin(), this->pass_statuses.end(), measurement_data.RangeStatus) == this->pass_statuses.end()) {
		char range_status[VL53L1_MAX_STRING_LENGTH];
		VL53L1_get_range_status_string(measurement_data.RangeStatus, range_status);
		RCLCPP_DEBUG(this->get_logger(), "Range measurement status is not valid: %s", range_status);
	} else {
		// Publish measurement
		this->range.range = measurement_data.RangeMilliMeter / 1000.0 + this->offset;
		this->range_pub->publish(this->range);
	}
}

int main(int argc, char **argv)
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<VL53L1X_Node>());
	rclcpp::shutdown();
	return 0;
}
